### PR TITLE
Artist screen

### PR DESCRIPTION
**Pull request by PWiley**
WIP ArtistCell

### CODE REVIEW

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+//  SoundsBinder
+//
+//  Created by Patrick Wiley on 02.02.22.
+//
+
+import UIKit
+
+final class ArtistsDataSource: NSObject, UICollectionViewDataSource, UICollectionViewDelegate {
+    
+    // MARK: - Private Properties
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:12]: missing space after mark

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        let index = indexPath.item
+        guard indexPath.item < catalogArtists.count else {
+            fatalError()
+        }
+        didSelectItemAtIndex?(index)
+    }
+
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:50]: useless line

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    // MARK: - UICollectionView Methods
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return catalogArtists.count
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let artistCell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath) as! ArtistCell
+        artistCell.configure(with: catalogArtists[indexPath.row])
+        artistCell.shadowDecorate()
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:39]: This is not the responsibility of the controller to build rendering. The cell should do it itself.

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        return artistCell
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        let index = indexPath.item
+        guard indexPath.item < catalogArtists.count else {
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:45]: the test should be done on viewModel side

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        }
+    }
+}
+
+extension ArtistsDataSource: UICollectionViewDelegateFlowLayout {
+    
+    func collectionView(_ collectionView: UICollectionView,
+                        layout collectionViewLayout: UICollectionViewLayout,
+                        sizeForItemAt indexPath: IndexPath) -> CGSize {
+
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:64]: useless line

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsRepository.swift [hunk 6]
```diff
@@ -1,27 +1,68 @@
...
-    func searchArtists(for name: String, callback: (Result<ArtistsResponse, ArtistsRepositoryError>) -> Void) {
-        
-    }
+   
+    // MARK: - Properties
+    
+    private let client: HTTPClientType
+    private let parser: JSONParserType
+    private let token = RequestCancellationToken()
+
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsRepository.swift:24]: useless line

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsRepository.swift [hunk 7]
```diff
@@ -1,27 +1,68 @@
...
+    private func handle(data: Data) -> [Artist] {
+        guard let result: Result<ArtistsResponse, ParserError> = try? self.parser.processCodableResponse(from: data) else {
+            return []
+        }
+        switch result {
+        case .success(let response):
+            return response.artists.map { Artist(item: $0) }
+        case .failure(let error):
+            //assertionFailure(error.localizedDescription)
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsRepository.swift:55]: dead code

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsViewController.swift [hunk 8]
```diff
@@ -40,37 +49,53 @@ final class ArtistsViewController: UIViewController {
...
+        view.addSubview(collectionView)
+        collectionView.snp.makeConstraints {
+            $0.top.bottom.leading.trailing.equalToSuperview()
+           
         }
     }
+    
+    private func bind(to viewModel: ArtistsViewModel) {
+        viewModel.items = { [weak self] items in
+            DispatchQueue.main.async {
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsViewController.swift:78]: Dispatch should be done on viewModel side

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        //image.translatesAutoresizingMaskIntoConstraints = false
+        return image
+    }()
+    
+    private lazy var label: UILabel = {
+        let label = UILabel()
+        label.text = ""
+        label.textColor = .white
+        label.textAlignment = .center
+        // label.adjustsFontSizeToFitWidth = true
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:29]: dead code

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        setupCellLayout()
+    }
+    
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+        setupCellLayout()
+    }
+    
+    // MARK: - Functions helper contraints
+
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:56]: useless line

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        
+        contentView.addSubview(label)
+        label.snp.makeConstraints {
+            $0.bottom.equalTo(imageView.snp.bottom)
+            $0.leftMargin.equalToSuperview()
+            $0.rightMargin.equalToSuperview()
+            $0.bottomMargin.equalToSuperview()
+        }
+
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:71]: useless line

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            $0.rightMargin.equalToSuperview()
+            $0.bottomMargin.equalToSuperview()
+        }
+        
+        
+    }
+    // MARK: - Setup
+    
+    func configure(with item: VisibleArtist) {
+
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:77]: useless line

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        if let imageUrl = item.pictureURLString {
+            repository?.downloadImage(for: URL(string: imageUrl)!, cancelledBy: RequestCancellationToken(), callback: { [weak self] data in
+                DispatchQueue.main.async {
+                    guard let data = data else { return }
+                    self?.imageView.image = UIImage(data: data)
+                }
+            })
+        }
+    }
+    func shadowDecorate() {
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:91]: should be private

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        contentView.layer.masksToBounds = true
+        
+        layer.shadowColor = UIColor.black.cgColor
+        layer.shadowOffset = CGSize(width: 0, height: 1.0)
+        layer.shadowRadius = 2.0
+        layer.shadowOpacity = 0.5
+        layer.masksToBounds = false
+        layer.shadowPath = UIBezierPath(roundedRect: bounds, cornerRadius: radius).cgPath
+        layer.cornerRadius = radius
+
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:105]: useless line

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Image/imageProvider.swift [hunk 15]
```diff
@@ -0,0 +1,40 @@
+////
+////  imageProvider.swift
+////  SoundsBinder
+////
+////  Created by Patrick Wiley on 02.02.22.
+////
+//
+//import Foundation
+//
+//final class ImageProvider {
```
Comment (Asavarkhul) [inline SoundsBinder/SoundsBinder/Sources/App/Artists/Image/imageProvider.swift:10]: Dead code

**Review by Asavarkhul — CHANGES_REQUESTED** [review]

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift (hunk 1 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:12]: ok

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift (hunk 2 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:50]: ok

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift (hunk 3 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:39]: ok

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift (hunk 4 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:45]: ok

File: SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift (hunk 5 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/ArtistsDataSource.swift:64]: ok

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift (hunk 12 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:77]: ok

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift (hunk 13 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:91]: not anymore used.

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift (hunk 14 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/Cell/ ArtistCell.swift:105]: not anymore used.

File: SoundsBinder/SoundsBinder/Sources/App/Artists/Image/imageProvider.swift (hunk 15 above)
Comment (PWiley) [reply SoundsBinder/SoundsBinder/Sources/App/Artists/Image/imageProvider.swift:10]: delete class

**Pull request closed without merging**